flutter/lib/src/span_frame_metrics_collector.dart

```diff
@@ -0,0 +1,212 @@
+  /// After the frames are calculated and stored in the span the span is removed from this list.
+  final activeSpans = <ISentrySpan>[];
+
+  bool get isTrackingPaused => _isTrackingPaused;
+  bool _isTrackingPaused = true;
+
+  bool get isTrackingRegistered => _isTrackingRegistered;
+  bool _isTrackingRegistered = false;
+
+  final _stopwatch = Stopwatch();
+
+  SpanFrameMetricsCollector(this.options,
+      {FrameCallbackHandler? frameCallbackHandler, SentryNative? native})
+      : _frameCallbackHandler =
+            frameCallbackHandler ?? DefaultFrameCallbackHandler(),
+        _native = native ?? SentryFlutter.native;
+
+  @override
+  void onSpanStarted(ISentrySpan span) {
+    if (span is NoOpSentrySpan || !options.enableFramesTracking) {
+      return;
+    }
+
+    activeSpans.add(span);
+    startFrameTracking();
+  }
+
+  @override
+  Future<void> onSpanFinished(ISentrySpan span, DateTime endTimestamp) async {
+    if (span is NoOpSentrySpan || !options.enableFramesTracking) {
+      return Future.value();
+    }
+
+    await recordSpanFrameMetrics(span, endTimestamp);
+
+    if (activeSpans.isEmpty) {
+      clear();
+    } else {
+      final oldestSpan = activeSpans.first;
+      frameDurations.removeWhere((key, value) {
+        return key.isBefore(oldestSpan.startTimestamp);
+      });
+    }
+  }
+
+  /// Calls [WidgetsBinding.instance.addPersistentFrameCallback] which cannot be unregistered
+  /// and exists for the duration of the application's lifetime.
+  ///
+  /// Stopping the frame tracking means setting [isTrackingPaused] is `true`
+  /// to prevent actions being done when the frame callback is triggered.
+  void startFrameTracking() {
+    _isTrackingPaused = false;
+
+    if (!_isTrackingRegistered) {
+      _frameCallbackHandler?.addPersistentFrameCallback(measureFrameDuration);
+      _isTrackingRegistered = true;
+    }
+  }
+
+  /// Records the duration of a single frame and stores it in [frameDurations].
+  ///
+  /// This method is called for each frame when frame tracking is active.
+  void measureFrameDuration(Duration duration) async {
+    if (_isTrackingPaused) return;
+
+    if (!_stopwatch.isRunning) {
+      _stopwatch.start();
+    }
+
+    await _frameCallbackHandler?.endOfFrame;
+
+    final frameDuration = _stopwatch.elapsedMilliseconds;
+    // ignore: invalid_use_of_internal_member
+    frameDurations[getUtcDateTime()] = frameDuration;
+
+    _stopwatch.reset();
+
+    if (_frameCallbackHandler?.hasScheduledFrame == true) {
+      _stopwatch.start();
+    }
+  }
+
+  Future<void> recordSpanFrameMetrics(
+      ISentrySpan span, DateTime endTimestamp) async {
+    final displayRefreshRate = await _native?.displayRefreshRate();
+    if (displayRefreshRate == null) {
+      options.logger(SentryLevel.warning,
+          'Display refresh rate is not available. Dropping the frame metrics');
+      clear();
+      return;
+    }
+
+    activeSpans.removeWhere(
```

Comment:
s: maybe we can change `activeSpans` to a set and here just do activeSpans.remove(span)?